File: Python/Distinct Subsequences.py
# Given a string S and a string T, count the number of distinct subsequences of T in S.

# A subsequence of a string is a new string which is formed from the original string by deleting some (can be none) of the characters without disturbing the relative positions of the remaining characters. (ie, "ACE" is a subsequence of "ABCDE" while "AEC" is not).

# Here is an example:
# S = "rabbbit", T = "rabbit"

# Return 3.


class Solution:
    # @return an integer
    def numDistinct(self, S, T):
		match = [0 for i in xrange(len(T)+1)]
		if len(S) < len(T):
			return 0
		match[0] = 1
		for i in range(1,len(S)+1):
			for j in range(len(T),0,-1):
				if S[i-1] == T[j-1]:
					match[j] = match[j] + match[j-1]
		return match[len(T)]

File: Java/Distinct Subsequences.java
class Distinct Subsequences{
	//dp strategy
	//table[i][j] means the number of distinct subsequence between S(0,i) and T(0,j);
	//table[i][j] = table[i-1][j] + (table[i-1][j-1] if s[i-1] == t[j-1])
	public int numDistinct(String S, String T) {
        int[][] table = new int[S.length()+1][T.length()+1];
        for(int i=0;i<=S.length();i++){
        	table[i][0] = 1;
        }
        for(int i=1;i<=S.length();i++){
        	for(int j=1;j<=T.length();j++){
        		if(S.charAt(i-1) == T.charAt(j-1)){
        			table[i][j] = table[i-1][j-1] + table[i-1][j];
        		}
        		else
        			table[i][j] = table[i-1][j];
        	}
        }
        return table[S.length()][T.length()];
    }
}